fastcrypto/src/groups/bls12381.rs

```diff
@@ -218,6 +258,24 @@ impl GroupElement for G2Element {
     }
 }
 
+impl FromHashedMessage for G2Element {
+    fn hash(msg: &[u8]) -> Self {
+        let mut res = blst_p2::default();
+        unsafe {
+            blst_hash_to_g2(
+                &mut res,
+                msg.as_ptr(),
+                msg.len(),
+                DST_G2.as_ptr(),
+                DST_G2.len(),
+                ptr::null(),
+                0,
+            );
+        }
+        Self::from(res)
+    }
+}
```

Comment:
OK, so this is the one that is actually confusing, let me explain how I think about inferring what you're trying to do here:
- in idiomatic Rust `FromXYZ` implemented for a type `Foo` indicate there's a mapping from `XYZ` to `Foo`. Here `hash` takes an arbitrary slice of `u8`, so it's unclear what `XYZ` is,
- `FromHashedMessage` would seem to indicate that you're reaching for `XYZ` == a field element, because there is indeed a setup phase in which we hash arbitrary messages to fixed-size field elements before trying to map them to points, [following the RFC-advised API](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#section-3). The fact that we do not quite have a field trait defined (hi @jonas-lj ) would explain why you'd be taking in bytes instead of a particular type (e.g. `blst_fp2` in the case of G2). OK, so I would have expected to find a call to `map_to_g2` in the implementation (or some variant, see the details of how to make `encode_to_g2` uniform, and here I find `hash_to_g2`, which itself contains a hash to the field. How are the arguments to `hash` "Hashed", exactly?
- Now I'm thinking you may be trying to model the idiosyncratic practice of the sui code base to sign and process the application-level hashes of messages instead of the messages themselves (i.e. we sign hashes of transactions rather than the transaction bytes). This practice is one I have found problematic for a while (#20) but it is pretty solidly entrenched in the Sui (and NW) code base today. The `FromHashedMessage` name would them mean "FromApplicationMessage", which make sense. 

Is this what you were trying to model @benr-ml ? @joyqvq did you understand this as well?

(side note: when it will come to implementing this, I'll advise looking at the RFC for the details of the difference between `blst_hash_to_g2` and `blst_encode_to_g2`, but that's a more involved discussion that's premature at this stage)